src/netif.c

```diff
@@ -1724,42 +1724,76 @@ static int remove_all_gateways6(struct netif *nb, const char *ifname)
 
 static int add_default_gateway(struct netif *nb, const char *ifname, const char *gateway_ip)
 {
-    debug("add_default_gateway %s %s", ifname, gateway_ip);
+  struct nlmsghdr *nlh = mnl_nlmsg_put_header(nb->nlbuf);
+  struct rtmsg    *rtm = mnl_nlmsg_put_extra_header(nlh, sizeof(struct rtmsg));;
 
-    struct rtentry route;
-    memset(&route, 0, sizeof(route));
+  int seq = nb->seq++;
+  int ret = 0;
 
-    struct sockaddr_in *addr = (struct sockaddr_in *)&route.rt_gateway;
-    memset(addr, 0, sizeof(struct sockaddr_in));
-    addr->sin_family = AF_INET;
-    if (inet_pton(AF_INET, gateway_ip, &addr->sin_addr) <= 0) {
-        debug("Bad IP address for the default gateway: %s", gateway_ip);
-        nb->last_error = EINVAL;
-        return -1;
+  memset(rtm, 0, sizeof(*rtm));
+
+  nlh->nlmsg_type = RTM_NEWROUTE;
+
+  nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_CREATE | NLM_F_ACK;
+  nlh->nlmsg_seq   = seq;
+
+  rtm->rtm_family   = AF_INET;
+  rtm->rtm_table    = RT_TABLE_MAIN;
+  rtm->rtm_scope    = RT_SCOPE_UNIVERSE;
+  rtm->rtm_protocol = RTPROT_BOOT;
+  rtm->rtm_type     = RTN_UNICAST;
+  rtm->rtm_flags    = RTNH_F_ONLINK;
+
+  {
+    struct sockaddr_in addr;
+    memset(&addr, 0, sizeof(struct sockaddr_in));
+    addr.sin_family = AF_INET;
+
+    if (inet_pton(AF_INET, gateway_ip, &addr.sin_addr) <= 0) {
+      debug("Bad IP address for the default gateway: %s", gateway_ip);
+      nb->last_error = EINVAL;
+      return -1;
     }
 
-    addr = (struct sockaddr_in*) &route.rt_dst;
-    memset(addr, 0, sizeof(struct sockaddr_in));
-    addr->sin_family = AF_INET;
-    addr->sin_addr.s_addr = INADDR_ANY;
+    mnl_attr_put_u32(nlh, RTA_GATEWAY, addr.sin_addr.s_addr);
+  }
 
-    addr = (struct sockaddr_in*) &route.rt_genmask;
-    memset(addr, 0, sizeof(struct sockaddr_in));
-    addr->sin_family = AF_INET;
-    addr->sin_addr.s_addr = INADDR_ANY;
+  if (ifname) {
+    int idx = ifname_to_index(nb, ifname);
 
-    route.rt_dev = (char *) ifname;
-    route.rt_flags = RTF_UP | RTF_GATEWAY;
-    route.rt_metric = 0;
+    if (!idx) {
+      debug("[%s %d %s]: No such device: '%s'\r\n", __FILE__, __LINE__, __FUNCTION__, ifname);
```

Comment:
want to keep debug print ?

Reply:
yup - it is compiled out via the ansi-c #debug pre-compiler's macro ;)